politeiawww/pi/error.go

```diff
@@ -0,0 +1,197 @@
+	var (
+		pluginID   = pde.ErrorReply.PluginID
+		errCode    = pde.ErrorReply.ErrorCode
+		errContext = pde.ErrorReply.ErrorContext
+	)
+	e := convertPDErrorCode(errCode)
+	switch {
+	case pluginID != "":
+		// politeiad plugin error. Log it and return a 400.
+		m := fmt.Sprintf("%v Plugin error: %v %v",
+			util.RemoteAddr(r), pluginID, errCode)
+		if errContext != "" {
+			m += fmt.Sprintf(": %v", errContext)
+		}
+		log.Infof(m)
+		util.RespondWithJSON(w, http.StatusBadRequest,
+			v1.PluginErrorReply{
+				PluginID:     pluginID,
+				ErrorCode:    errCode,
+				ErrorContext: errContext,
+			})
+		return
+
+	case e != v1.ErrorCodeInvalid:
+		// User error from politeiad that corresponds to a records user
+		// error. Log it and return a 400.
+		m := fmt.Sprintf("%v Records user error: %v %v",
+			util.RemoteAddr(r), e, v1.ErrorCodes[e])
+		if errContext != "" {
+			m += fmt.Sprintf(": %v", errContext)
+		}
+		log.Infof(m)
+		util.RespondWithJSON(w, http.StatusBadRequest,
+			v1.UserErrorReply{
+				ErrorCode:    e,
+				ErrorContext: errContext,
+			})
+		return
+
+	default:
+		// politeiad error does not correspond to a user error. Log it
+		// and return a 500.
+		ts := time.Now().Unix()
+		log.Errorf("%v %v %v %v Internal error %v: error code "+
+			"from politeiad: %v", util.RemoteAddr(r), r.Method, r.URL,
+			r.Proto, ts, errCode)
+
+		util.RespondWithJSON(w, http.StatusInternalServerError,
+			v1.ServerErrorReply{
+				ErrorCode: ts,
+			})
+		return
+	}
+}
+
+func convertPDErrorCode(errCode uint32) v1.ErrorCodeT {
+	// Any error statuses that are intentionally omitted means that
+	// politeiawww should 500.
+	switch pdv2.ErrorCodeT(errCode) {
+	case pdv2.ErrorCodeRequestPayloadInvalid:
+		// Intentionally omitted
+	case pdv2.ErrorCodeChallengeInvalid:
+		// Intentionally omitted
+	case pdv2.ErrorCodeMetadataStreamInvalid:
+		// Intentionally omitted
+	case pdv2.ErrorCodeMetadataStreamDuplicate:
+		// Intentionally omitted
+	case pdv2.ErrorCodeFilesEmpty:
+		// Intentionally omitted
+	case pdv2.ErrorCodeFileNameInvalid:
+		// Intentionally omitted
+	case pdv2.ErrorCodeFileNameDuplicate:
+		// Intentionally omitted
+	case pdv2.ErrorCodeFileDigestInvalid:
+		// Intentionally omitted
+	case pdv2.ErrorCodeFilePayloadInvalid:
+		// Intentionally omitted
+	case pdv2.ErrorCodeFileMIMETypeInvalid:
+		// Intentionally omitted
+	case pdv2.ErrorCodeFileMIMETypeUnsupported:
+		// Intentionally omitted
+	case pdv2.ErrorCodeTokenInvalid:
+		return v1.ErrorCodeRecordTokenInvalid
+	case pdv2.ErrorCodeRecordNotFound:
+		return v1.ErrorCodeRecordNotFound
+	case pdv2.ErrorCodeRecordLocked:
+		// Intentionally omitted
+	case pdv2.ErrorCodeNoRecordChanges:
+		// Intentionally omitted
+	case pdv2.ErrorCodeStatusChangeInvalid:
+		// Intentionally omitted
+	case pdv2.ErrorCodePluginIDInvalid:
+		// Intentionally omitted
+	case pdv2.ErrorCodePluginCmdInvalid:
+		// Intentionally omitted
+	}
+	return v1.ErrorCodeInvalid
```

Comment:
Aren't there suppose to be conversions for the `TokenInvalid` and `RecordNotFound`?